### PR TITLE
add TxOut decoration code for TxHistory

### PR DESCRIPTION
**Pull request by paolino**
- [x] I have added a TxOut decoration functionality to the transaction model. The function is pure and just a utility over the TxHistory store
- [x] Implemented property testing for the decoration

### Issue Number

ADP-1784

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs [hunk 1]
```diff
@@ -276,3 +283,54 @@ mkTxHistory txs = TxHistoryF $ fold $ do
...
     pure $ Map.singleton (TxId $ tx ^. #txId) relation
+
+type TxOutKey = (TxId, Word32)
+
+decorateWithTxOuts :: TxHistoryF 'Without -> TxHistoryF 'With
+decorateWithTxOuts (TxHistoryF w) =
+  let txouts :: Map TxOutKey (TxOut, [TxOutToken]) = Map.fromList $ do
+        TxRelationF {..} <- toList w
+        [(txOutputTxId &&& txOutputIndex $ txout, x) | x@(txout,_ ) <- outs]
+   in TxHistoryF $ fmap (solveTxOut txouts) w
```
Comment (HeinrichApfelmus) [inline lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs]:
```suggestion
    let txouts :: Map TxOutKey (TxOut, [TxOutToken])
        txouts = Map.fromList $ do
            TxRelationF{..} <- toList w
            [(txOutputTxId &&& txOutputIndex $ txout, x) | x@(txout,_ ) <- outs]
    in  TxHistoryF $ fmap (solveTxOut txouts) w
```

(Style

1. Indentation multiple of 4 spaces (with exception for `where` clause – 2 spaces)
2. Repeating the name `txouts` for type signature and definition seems worth the clarity.
3. I prefer to have no space between record constructor `TxRelationF` and curly braces `{}`; in this way, I can visually distinguish function arguments by spaces.

)

File: lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs [hunk 2]
```diff
@@ -276,3 +283,54 @@ mkTxHistory txs = TxHistoryF $ fold $ do
...
+            ins
+    , collateralIns =
+        decorateInputs
+            (txCollateralSourceTxId &&& txCollateralSourceIndex)
+            txOutMap
+            collateralIns
+    , outs = outs
+    , collateralOuts = collateralOuts
+    , withdrawals = withdrawals
+
```
Comment (HeinrichApfelmus) [inline lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs]:
```suggestion
```

File: lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs [hunk 3]
```diff
@@ -14,23 +19,172 @@ import Cardano.Wallet.DB.Fixtures
...
-spec = around withDBInMemory $ do
-    describe "Transactions store" $ do
-        it "respects store laws"
-            $ property . prop_StoreLaws
+spec = do
+    around withDBInMemory $ do
+        describe "Transactions store" $ do
+            it "respects store laws" $
+                property . prop_StoreLaws
+    describe "TxOut decoration" $ do
```
Comment (HeinrichApfelmus) [inline lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs:61]:
```suggestion

    describe "TxOut decoration" $ do
```

(Style: Empty line for better visual sectioning.)

File: lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs [hunk 4]
```diff
@@ -14,23 +19,172 @@ import Cardano.Wallet.DB.Fixtures
...
+            $ property prop_DecorateLinksTxInToTxOuts
+        it
+            "reports a transaction where collateral inputs point \
+            \to all other transactions output"
+            $ property prop_DecorateLinksTxCollateralsToTxOuts
+
+prop_DecorateIsInvertible :: [Tx] -> Bool
+prop_DecorateIsInvertible transactions =
+    let txh = mkTxHistory transactions
+     in undecorateFromTxOuts (decorateWithTxOuts txh) == txh
```
Comment (HeinrichApfelmus) [inline lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs:76]:
```suggestion
    in  undecorateFromTxOuts (decorateWithTxOuts txh) == txh
```
(Style: Seems to be the best of both worlds. 4 spaces indentation, _both_ before and after `in`. 🤔 )

File: lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs [hunk 5]
```diff
@@ -14,23 +19,172 @@ import Cardano.Wallet.DB.Fixtures
...
+    forAll transactionsGen $ \(idx, result, txouts) -> do
+        ins (result Map.! idx)
+            `shouldBe` [ WithTxOut txin (Just txout)
+                       | (_, (txin, txout)) <- txouts
+                       ]
+
+{- | We check that: given a set of random transactions and a guinea one for
+ which we set all collaterals to be pointing to all outputs of the others,
+ the decorated history contains a guinea transaction with the expected inputs
+-}
```
Comment (HeinrichApfelmus) [inline lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs]:
```suggestion
{- | We check that `decorateWithTxOuts` indeed decorates collateral inputs.

We do this by generating a set of random transactions, as well as a
"guinea pig" transaction, whose collateral inputs point to all outputs
of the other transactions. Then, we expect that decorating the history
will decorate all inputs of the "guinea pig" transaction.
-}
```

File: lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs [hunk 6]
```diff
@@ -14,23 +19,172 @@ import Cardano.Wallet.DB.Fixtures
...
+
+prop_DecorateIsInvertible :: [Tx] -> Bool
+prop_DecorateIsInvertible transactions =
+    let txh = mkTxHistory transactions
+     in undecorateFromTxOuts (decorateWithTxOuts txh) == txh
+
+{- | We check that given a set of random transactions and a guinea one for which
+ we set all inputs to be pointing to all outputs of the others,
+ the decorated history contains a guinea transaction with the expected inputs
+-}
```
Comment (HeinrichApfelmus) [inline lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs]:
```suggestion
{- | We check that `decorateWithTxOuts` indeed decorates transaction inputs.

We do this by generating a set of random transactions, as well as a
"guinea pig" transaction, whose inputs point to all outputs
of the other transactions. Then, we expect that decorating the history
will decorate all inputs of the "guinea pig" transaction.
-}
```

File: lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs [hunk 7]
```diff
@@ -14,23 +19,172 @@ import Cardano.Wallet.DB.Fixtures
...
+                    decorateWithTxOuts $
+                        mkTxHistory (guinea' : transactions)
+            pure (guineaId, result, txouts)
+    forAll transactionsGen $ \(idx, result, txouts) -> do
+        collateralIns (result Map.! idx)
+            `shouldBe` [ WithTxOut txin (Just txout)
+                       | (_, (txin, txout)) <- txouts
+                       ]
+
+generateLattice :: Int -> Double -> Gen (TxHistory, [W.TxIn])
```
Comment (HeinrichApfelmus) [inline lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs]:
```suggestion
-- | ???
generateLattice :: Int -> Double -> Gen (TxHistory, [W.TxIn])
```

Brief summary of what this function does? 🤔

**Review by HeinrichApfelmus — APPROVED** [review]
Looks good to me! Tidy code, I like that! 😊 Small style suggestions. For additional clarity, I have rephrased the comments on `prop_Decorate`* — here using a small "what. how." rephrasing trick.

Could you add a documenting comment on `generateLattice` as well? I don't fully understand it, but `generateLattice n f` seems to generate a history of `n` transactions, where each transaction contains a fraction `f` of the size parameter many inputs. 🤔

File: lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs (hunk 7 above)
Comment (paolino) [reply lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs]: I removed the generateLattice as @jonathanknowles has a shrinking solution for that in the pocket. I will postpone the idea for a while. generateLattice is just creating a valid sequence of transactions as opposed to a random bunch of them, following the utxo laws from a random genesis

File: lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs (hunk 6 above)
Comment (paolino) [reply lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs]: done

File: lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs (hunk 1 above)
Comment (paolino) [reply lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs]: done, with a bit of doubts

File: lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs (hunk 3 above)
Comment (paolino) [reply lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs:61]: done

**Comment by paolino** [comment]
> Looks good to me! Tidy code, I like that! blush Small style suggestions. For additional clarity, I have rephrased the comments on `prop_Decorate`* — here using a small "what. how." rephrasing trick.
> 
> Could you add a documenting comment on `generateLattice` as well? I don't fully understand it, but `generateLattice n f` seems to generate a history of `n` transactions, where each transaction contains a fraction `f` of the size parameter many inputs. thinking

GenerateLattice is suspended, see comments above.
Thanks for the docs, and the trick and the review.

**Pull request merged**